Update prom rw exporter

SdotG7 opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #933 

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Unit tests have been updated and should provide similar coverage of functionality from before. Some tests where replaced, others fixed to support changes to MetricExporter ABC & or new components from OT SDK
- Example directory for this exporter previously contained a complete example. I updated that to also provide similar coverage of the supported OT metric objects. It works as defined following the README instructions in there. No instruction update was required.

# Does This PR Require a Core Repo Change?
- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project (Ran linters, but may have missed some things)
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [x] Documentation has been updated

SdotG7 committed:
Proto & Tox updates

- add Tox commands to setup & test the exporter
- add the Prometheus Remote Write Proto files & update them.
- added a mini-script that wraps the protoc command to regen. them

SdotG7 committed:
Add some tests & test infra.

SdotG7 committed:
Update test cases

SdotG7 committed:
More fixes from Example run

- Fix some naming issues,add a regex to correct these
- Update setup.cfg versions
- Update the example app for newer sdk syntax

SdotG7 committed:
Ran linter

SdotG7 committed:
Version update

this would be released as part of 1.13.0

ocelotl committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Update setup.cfg to use >=3.7 instead of 3.8

SdotG7 committed:
Address automated checks

- Fix flake8 issues
- Correct spelling mistakes (except false positive 'ro' in docker
  compose)
- Updated Changelog. Not sure if this is worthy

srikanthccv committed:
Merge branch 'main' into update_prom_rw_exporter

srikanthccv committed:
Merge branch 'main' into update_prom_rw_exporter

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/README.rst:
```diff
@@ -0,0 +1,322 @@
+
+|Prometheus SDK pipelines|
+
+The Prometheus Remote Write Exporter is a "push" based exporter and only
+works with the OpenTelemetry `push controller`_. The controller
+periodically collects data and passes it to the exporter. This exporter
+then converts the data into `timeseries`_ and sends it to the Remote
+Write integrated backend through HTTP POST requests. The metrics
+collection datapath is shown below:
```
This comment is quite out of date now, should be `PeriodicExportingMetricReader`.

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/README.rst:
```diff
@@ -0,0 +1,322 @@
+   # The Meter is responsible for creating and recording metrics. Each meter has a unique name, which we set as the module's name here.
+   meter = metrics.get_meter(__name__)
+
+   exporter = PrometheusRemoteWriteMetricsExporter(endpoint="endpoint_here") # add other params as needed
+
+   metrics.get_meter_provider().start_pipeline(meter, exporter, 5)
```
Same comment as above, this example is outdated

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/README.rst:
```diff
@@ -0,0 +1,322 @@
+and key files in the ``tls_config`` parameter.
+
+Supported Aggregators
+---------------------
```
This is all very outdated,  maybe best to just remove it altogether

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/README.rst:
```diff
@@ -0,0 +1,322 @@
+
+   1 4 5
+
+Contributing
+------------
+
+If you would like to learn more about the exporter's structure and
+design decisions please view the design document below
+
+Design Doc
+~~~~~~~~~~
+
+`Design Document`_
+
+This document is stored elsewhere as it contains large images which will
+significantly increase the size of this repo.
```
I don't think these are needed

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/proto/generate-proto-py.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
```
can you add some directions for running this script? E.g. it looks like it needs to be run from this directory

Also instead of adding the proto files to this repo directly, could you make this script download and patch them? It's easier to avoid any copyright/licensing confusion and make it easier to update if new proto files are released.

You might be able to re-use some of this https://github.com/open-telemetry/opentelemetry-python/blob/main/scripts/proto_codegen.sh

aabmass replied:
That's what I had in mind, SGTM!

aabmass replied:
LGTM, can we just delete these proto files which are pulled in from the script after generation or add them to .gitignore?

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/setup.cfg:
```diff
@@ -0,0 +1,49 @@
+# Copyright The OpenTelemetry Authors
```
I think we've moved to pyproject.toml e.g. https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/exporter/opentelemetry-exporter-richconsole/pyproject.toml

@open-telemetry/opentelemetry-python-contrib-maintainers is there an easy way to migrate this?

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/setup.cfg:
```diff
@@ -0,0 +1,49 @@
+install_requires =
+    protobuf ~= 4.21
+    requests ~= 2.28
+    opentelemetry-api == 1.12.0
+    opentelemetry-sdk == 1.12.0
```
Any reason to pin?

```suggestion
    opentelemetry-api ~= 1.12.0
    opentelemetry-sdk ~= 1.12.0
```

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/__init__.py:
```diff
@@ -0,0 +1,393 @@
+
+logger = logging.getLogger(__name__)
+
+PROMETHEUS_NAME_REGEX = re.compile(r"[^\w:]")
+PROMETHEUS_LABEL_REGEX = re.compile(r"[^\w]")
```
Was this in the original code or just added? If restored, feel free to tackle in a follow up PR.

The regexes in the [Prometheus docs](https://prometheus.io/docs/concepts/data_model/#:~:text=The%20metric%20name%20specifies%20the%20general%20feature%20of%20a%20system%20that%20is%20measured%20(e.g.%20http_requests_total%20%2D%20the%20total%20number%20of%20HTTP%20requests%20received).%20It%20may%20contain%20ASCII%20letters%20and%20digits%2C%20as%20well%20as%20underscores%20and%20colons.%20It%20must%20match%20the%20regex%20%5Ba%2DzA%2DZ_%3A%5D%5Ba%2DzA%2DZ0%2D9_%3A%5D*.) is different, is that intentional?

aabmass replied:
Ah gotcha I was reading it wrong. Anchoring the \d at the beginning seems to do the trick:

```py
>>> re.sub(r"^\d|[^\w:]", "_", "4hellowor:ld5∂©∑")
'_hellowor:ld5___'
```

lmk if that works and if not feel free to just resolve this comment.

aabmass replied:
[One more requirement](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#prometheus-compatibility:~:text=the%20_%20character.-,Multiple%20consecutive%20_%20characters%20MUST%20be%20replaced%20with%20a%20single%20_%20character.,-The%20Unit%20of). Feel free to open an issue and either do in a follow up PR or I can mark it up for grabs.

Seefooo replied:
That link isn't working for me for some reason, but I THINK you're pointing at the consecutive "_" being replaced? I took a whack at that and got it working, but it does clobber the `__name__` label with how things are currently structured. Its fixable, but probably more work than is worth it in this ticket. I will open an issue to avoid bogging this down.

Seefooo replied:
Issued opened here:
https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1389

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/__init__.py:
```diff
@@ -0,0 +1,393 @@
+    def __init__(
+        self,
+        endpoint: str,
+        basic_auth: Dict = None,
```
Were these types annotations in this PR or from the original?

aabmass replied:
Cool, then no need to address in this PR

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/__init__.py:
```diff
@@ -0,0 +1,393 @@
+    def force_flush(self, timeout_millis: float = 10_000) -> bool:
+        return True
+
+    def shutdown(self) -> None:
+        pass
```
This doesn't implement the ABC correctly https://github.com/open-telemetry/opentelemetry-python/blob/6f6f8d19babbb166816cb1d49b59d7985cf9ed11/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/export/__init__.py#L116-L121

aabmass reviewed (commented):
Thanks for PR! Just leaving a few comments but overall looks great 😃

SdotG7 committed:
Updates from review

- Update to pyproject.toml & float deps
- Update README.rst
- Updates to automate pulling & generating protobuf files
- Add README for protobuf generation

SdotG7 committed:
Fix the shutdown method

aabmass committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Remove extra proto files & update regex

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/examples/sampleapp.py:
```diff
@@ -0,0 +1,100 @@
+import logging
```
I think we need the copyright header here

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/pyproject.toml:
```diff
@@ -0,0 +1,51 @@
+dependencies = [
+  "protobuf ~= 4.21",
+  "requests ~= 2.28",
+  "opentelemetry-api ~= 1.13.0",
+  "opentelemetry-sdk ~= 1.13.0",
```
I believe `~= 1.13.0` is equivalent to `>= 1.13.0, == 1.13.0`. But it should be compatible with other minor versions right? Going off the template:

https://github.com/open-telemetry/opentelemetry-python-contrib/blob/9beb6b29b08a0c5333cb01139134eab03bbc11c8/_template/pyproject.toml#L31

does this work?

```suggestion
  "opentelemetry-api ~= 1.12",
  "opentelemetry-sdk ~= 1.12",
```

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/version.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__version__ = "1.13.0"
```
Should be `__version__ = "0.34b0"` like other packages in this repo since we can't guarantee stability until the spec is stable

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/tests/conftest.py:
```diff
@@ -0,0 +1,63 @@
+
+import pytest
+
+import opentelemetry.test.metrictestutil as metric_util  # import _generate_gauge, _generate_sum
```
```suggestion
import opentelemetry.test.metrictestutil as metric_util
```

aabmass commented on exporter/opentelemetry-exporter-prometheus-remote-write/src/opentelemetry/exporter/prometheus_remote_write/__init__.py:
```diff
@@ -0,0 +1,405 @@
+
+    def _parse_histogram_data_point(self, data_point, name):
+
+        # if (len(data_point.explicit_bounds)+1) != len(data_point.bucket_counts):
+        #    raise ValueError("Number of buckets must be 1 more than the explicit bounds!")
+
```
```suggestion
```

aabmass reviewed (approved):
LGTM with just a few changes. Thanks again for contributing :)

SdotG7 committed:
More updates from PR review

Seefooo commented on pyproject.toml:
```diff
@@ -6,6 +6,7 @@ exclude = '''
   | \.tox
   | venv
   | build
+  | gen
```
Note I made this change. This makes black's behavior consistent with flake8 & isort. otherwise it does reformat the generated protobuf files

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

srikanthccv committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Undo adding 'gen' to ignore list for black

- Was causing issues with other tools. I just ran our formatting tool on
  the generated files to avoid issues.
- add a gitignore here to ensure our mimicked import setup doesn't get
  committed.

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

SdotG7 committed:
Fixes from pylint output

- Cannot fix the syntax failures on 'examples/*' directory tho...

SdotG7 committed:
Move to `examples` dir to `example`

- avoids the pylint issue. This is probably better anyway as the example
  is focused on setting up an environment to test exporting, rather than
  providing multiple examples.

SdotG7 committed:
Cleanup README

SdotG7 committed:
Merge branch 'main' into update_prom_rw_exporter

srikanthccv commented:
Thanks!
